modules/stream-shredder/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/shredder/stream/sinks/generic/Record.scala

```diff
@@ -0,0 +1,110 @@
+package com.snowplowanalytics.snowplow.rdbloader.shredder.stream.sinks.generic
+
+import scala.annotation.tailrec
+import scala.concurrent.duration._
+
+import cats.Applicative
+import cats.implicits._
+
+import cats.effect.{Timer, Concurrent}
+
+import fs2.{Pipe, RaiseThrowable, Stream, Pure, Pull}
+
+/**
+ * A generic stream type, either holding data with associated window
+ * or denoting that window is over. Latter is necessary for downstream
+ * sinks to trigger actions without elements with next window
+ */
+sealed trait Record[F[_], W, A] extends Product with Serializable {
+  def window: W
+
+  def map[B](f: A => B): Record[F, W, B] =
+    this match {
+      case Record.Data(window, checkpoint, item) => Record.Data[F, W, B](window, checkpoint, f(item))
+      case Record.EndWindow(window, next, checkpoint) => Record.EndWindow(window, next, checkpoint)
+    }
+
+  def traverse[G[_]: Applicative, B](f: A => G[B]): G[Record[F, W, B]] =
+    this match {
+      case Record.Data(window, checkpoint, item) => f(item).map(b => Record.Data(window, checkpoint, b))
+      case Record.EndWindow(window, next, checkpoint) => Applicative[G].pure(Record.EndWindow(window, next, checkpoint))
+    }
+}
+
+object Record {
+
+  /**
+   * Actual windowed datum
+   * Not every `Data` can be checkpointed because a `Data` can be a middle of
+   * original record e.g. if it's one of multiple contexts
+   */
+  final case class Data[F[_], W, A](window: W, checkpoint: Option[F[Unit]], item: A) extends Record[F, W, A]
+  /** It belongs to `window`, but knows about `next` window */
+  final case class EndWindow[F[_], W, A](window: W, next: W, checkpoint: F[Unit]) extends Record[F, W, A]
+
+  def windowed[F[_]: Concurrent: Timer, W, A](getWindow: F[W]): Pipe[F, (A, F[Unit]), Record[F, W, A]] = {
+    in =>
+      val merged = Stream
+        .fixedRate(1.second)
+        .evalMap(_ => getWindow)
+        .either(in)
+
+      val initialWindow = Stream.eval(getWindow.map(_.asLeft))  // To prevent IllegalStateException
+
+      (initialWindow ++ merged).through(fromEither)
+  }
+
+  /**
+   * This `Pipe` restructures stream of `Either[W, A]` into a stream such that
+   * a sequence of `Record.Data` was always followed by a `Record.EndWindow`
+   * Must be used in conjunction with `windowed` because it ensures that head
+   * is always `Left`
+   */
+  private def fromEither[F[_]: RaiseThrowable: Applicative, W, A]: Pipe[F, Either[W, (A, F[Unit])], Record[F, W, A]] = {
```

Comment:
For fun, I re-wrote fromEither to use scan instead of using Pulls

```scala
  case class LocalState[F[_], W](w: W, checkpoint: F[Unit])

  private def fromEither[F[_], W, A](implicit AE: ApplicativeError[F, Throwable]): Pipe[F, Either[W, (A, F[Unit])], Record[F, W, A]] =
    in =>
      stateMap[F, Either[W, (A, F[Unit])], Record[F, W, A], Option[LocalState[F, W]]](in, Option.empty[LocalState[F, W]]) {
        case (None, Left(w)) =>
          AE.pure(Some(LocalState(w, AE.unit)) -> None)
        case (None, Right(_)) =>
          AE.raiseError(new IllegalStateException("windowing hasn't started"))
        case (Some(LocalState(lastWindow, _)), Left(w)) if w == lastWindow =>
          AE.pure(Some(LocalState(w, AE.unit)) -> None)
        case (Some(LocalState(lastWindow, lastChk)), Left(w)) =>
          AE.pure(Some(LocalState(w, AE.unit)) -> Some(Record.EndWindow(lastWindow, w, lastChk)))
        case (Some(LocalState(lastWindow, _)), Right((a, chk))) =>
          AE.pure(Some(LocalState(lastWindow, chk)) -> Some(Record.Data(lastWindow, Some(chk), a)))
      }

  private def stateMap[F[_], In, Out, S](stream: Stream[F, In], initial: S)(f: (S, In) => F[(S, Option[Out])]): Stream[F, Out] =
    stream.evalScan((initial, Option.empty[Out])) {
      case ((state, _), next) => f(state, next)
    }.collect {
      case (_, Some(next)) => next
    }
```

I haven't tested it and there may be errors, but it compiles.

My point is, there is always a way to write things without Pulls.  It is up for debate whose version is easier to read and to understand!! I *marginally* prefer my version for clarity, and I would have more trust that fs2 implements the scan with an efficient use of chunks and pulls.

Reply:
You're right, I probably got carried away with `Pull` after initial `Partitioned` experience.